### PR TITLE
[presto-native] Reuse driver instance for all scheduled splits

### PR DESCRIPTION
**Pull request by vermapratyush**
In presto-on-spark, a new driver runner instance is created for each split scheduled to run on a worker. For presto-on-spark native, this results in each split being processed by its own cpp instance, serially, thus poor query latency.
In the proposed solution, specifically for presto-on-spark native, we schedule all the splits on a single driver runner instance and thus single cpp instance.

For non-native flow, this change is a no-op.

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkTaskExecution.java [hunk 1]
```diff
@@ -224,10 +225,17 @@ private synchronized void scheduleTableScanSource(DriverSplitRunnerFactory facto
 
         // Enqueue driver runners with split lifecycle for this plan node and driver life cycle combination.
         ImmutableList.Builder<DriverSplitRunner> runners = ImmutableList.builder();
-        for (ScheduledSplit scheduledSplit : splits) {
-            // create a new driver for the split
-            runners.add(factory.createDriverRunner(scheduledSplit));
+        // In case of native execution we will pass all splits into one driver runner
```
Comment (mbasmanova) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkTaskExecution.java:228]: nit: Consider using active voice in all comments.

```
// For native execution, process all splits in a single driver.

// For Java execution, process each split in a separate driver.
```

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkTaskExecution.java [hunk 2]
```diff
@@ -224,10 +225,17 @@ private synchronized void scheduleTableScanSource(DriverSplitRunnerFactory facto
...
+        if (isNativeExecutionEnabled(this.taskContext.getSession())) {
+            runners.add(factory.createDriverRunner(splits));
         }
+        else { // While for non-native, we will create driver runner for each split
+            for (ScheduledSplit scheduledSplit : splits) {
+                // create a new driver for the split
```
Comment (mbasmanova) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkTaskExecution.java:234]: This comment is redundant. Let's remove.

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkTaskExecution.java [hunk 3]
```diff
@@ -339,25 +347,24 @@ private DriverSplitRunnerFactory(DriverFactory driverFactory, boolean partitione
...
-            long splitWeight = partitionedSplit == null ? 0 : partitionedSplit.getSplit().getSplitWeight().getRawValue();
-            DriverContext driverContext = pipelineContext.addDriverContext(splitWeight, Lifespan.taskWide(), driverFactory.getFragmentResultCacheContext());
-            return new DriverSplitRunner(this, driverContext, partitionedSplit);
+            // splitWeight can be 0 as we don't load balance the executor based on their load average
+            DriverContext driverContext = pipelineContext.addDriverContext(0, Lifespan.taskWide(), driverFactory.getFragmentResultCacheContext());
+            return new DriverSplitRunner(this, driverContext, scheduledSplitList);
         }
 
-        public Driver createDriver(DriverContext driverContext, @Nullable ScheduledSplit partitionedSplit)
+        public Driver createDriver(DriverContext driverContext, @Nullable List<ScheduledSplit> partitionedSplitList)
```
Comment (mbasmanova) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkTaskExecution.java:360]: same

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/operator/NativeExecutionOperator.java [hunk 4]
```diff
@@ -97,6 +99,8 @@
     private TaskSource taskSource;
     private boolean finished;
 
+    private List<ScheduledSplit> splitList = new ArrayList<>();
```
Comment (mbasmanova) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/execution/operator/NativeExecutionOperator.java:102]: splitList -> splits

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/operator/NativeExecutionOperator.java [hunk 5]
```diff
@@ -243,13 +247,11 @@ public PlanNodeId getSourceId()
...
-        checkState(this.taskSource == null, "NativeEngine operator split already set");
 
         if (finished) {
             return Optional::empty;
         }
-
-        this.taskSource = new TaskSource(split.getPlanNodeId(), ImmutableSet.of(split), true);
+        splitList.add(split);
 
         Object splitInfo = split.getSplit().getInfo();
```
Comment (mbasmanova) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/execution/operator/NativeExecutionOperator.java:256]: Is this logic correct? Shouldn't the setInfoSupplier be based on all splits, not the last one?

**Review by mbasmanova — COMMENTED** [review]
@vermapratyush Thank you for working on this. I haven't read the tests yet, but here are some comments and questions about the code. 

Please, update commit message. See current one at https://github.com/prestodb/presto/pull/19201/commits

File: presto-spark-base/src/test/java/com/facebook/presto/spark/task/TestNonNativePrestoSparkTaskExecution.java [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+    }
+
+    @Test
+    public void testNonNativeSourceNodeDriverSplitRunnerInstanceCount()
+    {
+        TaskContext nonNativeTaskContext = TestingTaskContext.createTaskContext(taskNotificationExecutor, scheduledExecutor, nonNativeTestSession, new DataSize(2, GIGABYTE));
+        taskExecutor.start();
+        PrestoSparkTaskExecution taskExecution = new PrestoSparkTaskExecution(taskStateMachine, nonNativeTaskContext, localExecutionPlan, taskExecutor, TaskTestUtils.createTestSplitMonitor(), taskNotificationExecutor, scheduledExecutor, false);
+        taskExecution.start(ImmutableList.of(new TaskSource(TABLE_SCAN_NODE_ID, ImmutableSet.of(split1, split2, split3), true)));
+        assertEquals(nonNativeTaskContext.getPipelineContexts().get(0).getPipelineStats().getDrivers().size(), 3);
```
Comment (mbasmanova) [inline presto-spark-base/src/test/java/com/facebook/presto/spark/task/TestNonNativePrestoSparkTaskExecution.java:105]: Is this the only difference between the 2 tests (in addition to setting NATIVE_EXECUTION_ENABLED to true or false)? 

If that's the case, perhaps, combine these 2 tests into one to avoid so much copy paste.

File: presto-spark-base/src/test/java/com/facebook/presto/spark/task/UnsupportedOutputBuffer.java [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+import com.facebook.presto.execution.buffer.OutputBufferInfo;
+import com.facebook.presto.execution.buffer.OutputBuffers;
+import com.facebook.presto.spi.page.SerializedPage;
+import com.google.common.util.concurrent.ListenableFuture;
+import io.airlift.units.DataSize;
+
+import java.util.List;
+import java.util.function.Consumer;
+
+class UnsupportedOutputBuffer
```
Comment (mbasmanova) [inline presto-spark-base/src/test/java/com/facebook/presto/spark/task/UnsupportedOutputBuffer.java:30]: This looks very similar to com.facebook.presto.sql.planner.TestLocalExecutionPlanner.TestingOutputBuffer

Can we reuse the code?

File: presto-spark-base/src/test/java/com/facebook/presto/spark/task/UnsupportedRemoteSourceFactory.java [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+import com.facebook.presto.Session;
+import com.facebook.presto.common.block.SortOrder;
+import com.facebook.presto.common.type.Type;
+import com.facebook.presto.operator.SourceOperatorFactory;
+import com.facebook.presto.spi.plan.PlanNodeId;
+import com.facebook.presto.sql.planner.RemoteSourceFactory;
+
+import java.util.List;
+
+public class UnsupportedRemoteSourceFactory
```
Comment (mbasmanova) [inline presto-spark-base/src/test/java/com/facebook/presto/spark/task/UnsupportedRemoteSourceFactory.java:25]: Same: com.facebook.presto.sql.planner.TestLocalExecutionPlanner.TestingRemoteSourceFactory

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/operator/NativeExecutionOperator.java (hunk 5 above)
Comment (vermapratyush) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/execution/operator/NativeExecutionOperator.java:256]: This function `addSplit` is passed 1 split at a time, so `operatorContext.setInfoSupplier` will be called once for each split anyway.

Comment (mbasmanova) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/execution/operator/NativeExecutionOperator.java:256]: This is used in com.facebook.presto.operator.OperatorContext#getOperatorStats and will produce incomplete stats, no?

File: presto-spark-base/src/test/java/com/facebook/presto/spark/task/TestPrestoSparkTaskExecution.java [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+{
+    Session nativeTestSession;
+    Session nonNativeTestSession;
+    LocalExecutionPlanner.LocalExecutionPlan localExecutionPlan;
+    TaskExecutor taskExecutor;
+    PlanFragment planFragment = createPlanFragment();
+    ExecutorService taskNotificationExecutor;
+    ScheduledExecutorService scheduledExecutor;
+    TaskStateMachine taskStateMachine;
+    ScheduledSplit split1 = new ScheduledSplit(1, TABLE_SCAN_NODE_ID, new Split(new ConnectorId("test"), TestingTransactionHandle.create(), TestingSplit.createLocalSplit()));
```
Comment (mbasmanova) [inline presto-spark-base/src/test/java/com/facebook/presto/spark/task/TestPrestoSparkTaskExecution.java:71]: Do we need all these split1, split2,... variables? Can we create a single List<ScheduledSplit> splits = ImmutableList.of(....) and use in both tests? Any particular reason Java test uses splits 1, 2, 3 and Native uses splits 4, 5?

File: presto-spark-base/src/test/java/com/facebook/presto/spark/task/TestPrestoSparkTaskExecution.java [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+        taskStateMachine.finished();
+        taskExecutor.stop();
+        scheduledExecutor.shutdown();
+        taskNotificationExecutor.shutdown();
+    }
+
+    @Test
+    public void testNativeSourceNodeDriverSplitRunnerInstanceCount()
+    {
+        TaskContext nativeTaskContext = TestingTaskContext.createTaskContext(taskNotificationExecutor, scheduledExecutor, nativeTestSession, new DataSize(2, GIGABYTE));
```
Comment (mbasmanova) [inline presto-spark-base/src/test/java/com/facebook/presto/spark/task/TestPrestoSparkTaskExecution.java:117]: Would it be possible to reduce copy-paste further?

```
@Test
public void testNativeDriverCount() 
{
   testDriverCount(true);
}

@Test
public void testJavaDriverCount() 
{
   testDriverCount(false)
}
```

**Review by mbasmanova — COMMENTED** [review]
@vermapratyush Nice refactoring. Thanks.

Let's split the changes into 2 commits and provide nice commit messages for each.

- First commit is to refactor TestLocalExecutionPlanner.
- Second commit is to make changes to native code path.

File: presto-spark-base/src/test/java/com/facebook/presto/spark/task/TestPrestoSparkTaskExecution.java (hunk 9 above)
Comment (vermapratyush) [reply presto-spark-base/src/test/java/com/facebook/presto/spark/task/TestPrestoSparkTaskExecution.java:71]: Actually, they don't require different splits anymore. 
I was initially having trouble having both of them in same test as I was not able to trigger a proper tearDown after each method. This is fixed now.

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/operator/NativeExecutionOperator.java (hunk 5 above)
Comment (vermapratyush) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/execution/operator/NativeExecutionOperator.java:256]: As discussed I have added a `TODO` for this section. #19279

File: presto-spark-base/src/test/java/com/facebook/presto/spark/task/TestPrestoSparkTaskExecution.java [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+    public void tearDown()
+    {
+        taskStateMachine.finished();
+        taskExecutor.stop();
+        scheduledExecutor.shutdown();
+        taskNotificationExecutor.shutdown();
+    }
+
+    @Test
+    public void testNativeSourceNodeDriverSplitRunnerInstanceCount()
```
Comment (mbasmanova) [inline presto-spark-base/src/test/java/com/facebook/presto/spark/task/TestPrestoSparkTaskExecution.java:118]: These names are super long and hard to read. Any particular reason not to shorten them to something like testNativeDriverCount and testJavaDriverCount?

**Review by mbasmanova — APPROVED** [review]
@vermapratyush Looks great. Thank you for iterating on this change. Please, update commit messages to wrap each line to 80 characters and mention a TODO.

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/operator/NativeExecutionOperator.java [hunk 12]
```diff
@@ -268,6 +270,7 @@ else if (splitInfo != null) {
     @Override
     public void noMoreSplits()
     {
+        this.taskSource = new TaskSource(splits.get(0).getPlanNodeId(), ImmutableSet.copyOf(splits), true);
```
Comment (miaoever) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/execution/operator/NativeExecutionOperator.java:273]: This doesn't look right to me.  `addSplit` will be called multiple times with splits associating with different source operators (think of join query with two TableScans), if you only use the planNodeId from the first split, then all splits will only go to the first operator which is not correct.

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/operator/NativeExecutionOperator.java (hunk 5 above)
Comment (vermapratyush) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/execution/operator/NativeExecutionOperator.java:256]: I have added this in line 259

File: presto-main/src/test/java/com/facebook/presto/sql/planner/TestLocalExecutionPlanner.java [hunk 13]
```diff
@@ -233,162 +219,6 @@ private LocalExecutionPlan getLocalExecutionPlan(Session session)
                 new TableWriteInfo(Optional.empty(), Optional.empty(), Optional.empty()));
     }
 
-    private static class TestingOutputBuffer
```
Comment (miaoever) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/TestLocalExecutionPlanner.java:236]: Is this change relevant to this PR?

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkTaskExecution.java [hunk 14]
```diff
@@ -224,10 +228,16 @@ private synchronized void scheduleTableScanSource(DriverSplitRunnerFactory facto
...
-        for (ScheduledSplit scheduledSplit : splits) {
-            // create a new driver for the split
-            runners.add(factory.createDriverRunner(scheduledSplit));
+        // For native execution, process all splits in a single driver.
+        if (isNativeExecutionEnabled(this.taskContext.getSession())) {
+            runners.add(factory.createDriverRunner(splits));
         }
+        else { // For Java execution, process each split in a separate driver.
+            for (ScheduledSplit scheduledSplit : splits) {
+                runners.add(factory.createDriverRunner(Lists.newArrayList(scheduledSplit)));
```
Comment (miaoever) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkTaskExecution.java:237]:
```
                runners.add(factory.createDriverRunner(ImmutableList.of(scheduledSplit)));

```

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkTaskExecution.java [hunk 15]
```diff
@@ -339,25 +349,26 @@ private DriverSplitRunnerFactory(DriverFactory driverFactory, boolean partitione
...
 
-        public Driver createDriver(DriverContext driverContext, @Nullable ScheduledSplit partitionedSplit)
+        public Driver createDriver(DriverContext driverContext, @Nullable List<ScheduledSplit> scheduledSplits)
         {
             Driver driver = driverFactory.createDriver(driverContext);
-            if (partitionedSplit != null) {
+            if (scheduledSplits != null && scheduledSplits.size() > 0) {
                 boolean isNativeExecutionEnabled = isNativeExecutionEnabled(driver.getDriverContext().getSession());
-                PlanNodeId sourceNodeId = isNativeExecutionEnabled ? driver.getSourceId().get() : partitionedSplit.getPlanNodeId();
+                PlanNodeId sourceNodeId = isNativeExecutionEnabled ? driver.getSourceId().get() : Iterables.getOnlyElement(scheduledSplits).getPlanNodeId();
```
Comment (miaoever) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkTaskExecution.java:369]: `Iterables.getOnlyElement` will throw exception if the list contains multiple element, should we do a pre-check?

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkTaskExecution.java [hunk 16]
```diff
@@ -339,25 +349,26 @@ private DriverSplitRunnerFactory(DriverFactory driverFactory, boolean partitione
             this.pipelineContext = taskContext.addPipelineContext(driverFactory.getPipelineId(), driverFactory.isInputDriver(), driverFactory.isOutputDriver(), partitioned);
         }
 
-        public DriverSplitRunner createDriverRunner(@Nullable ScheduledSplit partitionedSplit)
+        public DriverSplitRunner createDriverRunner(@Nullable List<ScheduledSplit> scheduledSplits)
```
Comment (miaoever) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkTaskExecution.java:352]: is the list still @Nullable ?

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkTaskExecution.java [hunk 17]
```diff
@@ -418,16 +429,16 @@ public void splitsAdded(int count, long weightSum)
...
         @GuardedBy("this")
         private Driver driver;
 
-        private DriverSplitRunner(DriverSplitRunnerFactory driverSplitRunnerFactory, DriverContext driverContext, @Nullable ScheduledSplit partitionedSplit)
+        private DriverSplitRunner(DriverSplitRunnerFactory driverSplitRunnerFactory, DriverContext driverContext, @Nullable List<ScheduledSplit> scheduledSplits)
         {
             this.driverSplitRunnerFactory = requireNonNull(driverSplitRunnerFactory, "driverFactory is null");
             this.driverContext = requireNonNull(driverContext, "driverContext is null");
-            this.partitionedSplit = partitionedSplit;
+            this.scheduledSplits = scheduledSplits;
```
Comment (miaoever) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkTaskExecution.java:441]: do a `requireNonNull ` check for the `scheduledSplits `?

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkTaskExecution.java [hunk 18]
```diff
@@ -339,25 +349,26 @@ private DriverSplitRunnerFactory(DriverFactory driverFactory, boolean partitione
...
             pendingCreation.incrementAndGet();
             // create driver context immediately so the driver existence is recorded in the stats
-            // the number of drivers is used to balance work across nodes
-            long splitWeight = partitionedSplit == null ? 0 : partitionedSplit.getSplit().getSplitWeight().getRawValue();
-            DriverContext driverContext = pipelineContext.addDriverContext(splitWeight, Lifespan.taskWide(), driverFactory.getFragmentResultCacheContext());
-            return new DriverSplitRunner(this, driverContext, partitionedSplit);
+            // splitWeight can be 0 as we don't load balance the executor based on their load average
+            DriverContext driverContext = pipelineContext.addDriverContext(0, Lifespan.taskWide(), driverFactory.getFragmentResultCacheContext());
+            DriverSplitRunner r = new DriverSplitRunner(this, driverContext, scheduledSplits);
+            r.getInfo();
```
Comment (miaoever) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkTaskExecution.java:360]: Why do we call `r.getInfo()` here?

**Review by miaoever — CHANGES_REQUESTED** [review]
Thanks for working on this issue @vermapratyush, I left a few comments.

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/operator/NativeExecutionOperator.java (hunk 12 above)
Comment (mbasmanova) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/execution/operator/NativeExecutionOperator.java:274]: This is a good point.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/TestLocalExecutionPlanner.java (hunk 13 above)
Comment (vermapratyush) [reply presto-main/src/test/java/com/facebook/presto/sql/planner/TestLocalExecutionPlanner.java:236]: I have moved it to public scope, so that I can re-use it in a test.

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkTaskExecution.java (hunk 16 above)
Comment (vermapratyush) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkTaskExecution.java:352]: yes, still nullable

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkTaskExecution.java (hunk 15 above)
Comment (vermapratyush) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkTaskExecution.java:369]: If it is not the only element in a non-native execution, we likely are doing something really wrong exampke: not evaluating all splits that were assigned.
So I think the exception is needed.

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/operator/NativeExecutionOperator.java (hunk 12 above)
Comment (mbasmanova) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/execution/operator/NativeExecutionOperator.java:274]: The existing code has this check that will fail if there is > 1 split (e.g. for a join query):

```
checkState(this.taskSource == null, "NativeEngine operator split already set");
```

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkTaskExecution.java (hunk 17 above)
Comment (vermapratyush) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkTaskExecution.java:441]: this is a nullable field.

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkTaskExecution.java (hunk 15 above)
Comment (miaoever) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkTaskExecution.java:367]: Yeah, I mean we should explicitly check the list only containing one element and throw a more meaningful exception msg if not (say `native execution requires only one scheduledSplits but [%d] were found.`) to provide more debugging context.

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/operator/NativeExecutionOperator.java (hunk 12 above)
Comment (vermapratyush) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/execution/operator/NativeExecutionOperator.java:274]: I am wondering if it is ok for `this.taskSource` to contain a list of TaskSource some of which comes from different TableScan?
If so, then we can change this to:
```
this.taskSource = splits.stream().map(split ->
                new TaskSource(split.getPlanNodeId(), ImmutableSet.of(split), true))
                .collect(Collectors.toList());
```

**Comment by tanjialiang** [comment]
Thanks @vermapratyush for working on this.

**Review by tanjialiang — CHANGES_REQUESTED** [review]
@vermapratyush This PR is breaking existing Spark native tests as well. There is a NullPointerException thrown when adding splits. It looks like taskSource list is not initialized. Can we fix this? And make sure things pass before landing. Thanks!

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkTaskExecution.java (hunk 16 above)
Comment (vermapratyush) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkTaskExecution.java:352]: Line 199 is passing null value.

**Comment by tanjialiang** [comment]
Thanks it looks good

**Review by tanjialiang — APPROVED** [review]

**Review by miaoever — APPROVED** [review]
LGTM, Thanks for the fix.

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/operator/NativeExecutionOperator.java [hunk 19]
```diff
@@ -243,18 +247,18 @@ public PlanNodeId getSourceId()
...
     {
         requireNonNull(split, "split is null");
-        checkState(this.taskSource == null, "NativeEngine operator split already set");
 
         if (finished) {
             return Optional::empty;
         }
-
-        this.taskSource = new TaskSource(split.getPlanNodeId(), ImmutableSet.of(split), true);
+        splits.add(split);
```
Comment (miaoever) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/execution/operator/NativeExecutionOperator.java:254]: Not needed right?

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/operator/NativeExecutionOperator.java [hunk 20]
```diff
@@ -94,9 +96,11 @@
     private NativeExecutionProcess process;
     private NativeExecutionTask task;
     private CompletableFuture<Void> taskStatusFuture;
-    private TaskSource taskSource;
+    private List<TaskSource> taskSource = new ArrayList<>();
     private boolean finished;
 
+    private List<ScheduledSplit> splits = new ArrayList<>();
```
Comment (miaoever) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/execution/operator/NativeExecutionOperator.java:102]: not needed.
